src/Http/WebUtilities/src/FileBufferingWriteStream.cs

```diff
@@ -0,0 +1,300 @@
+            => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            ThrowArgumentException(buffer, offset, count);
+            ThrowIfDisposed();
+
+            if (_bufferLimit.HasValue && _bufferLimit - BufferedLength < count)
+            {
+                Dispose();
+                throw new IOException("Buffer limit exceeded.");
+            }
+
+            // Allow buffering in memory if we're below the memory threshold once the current buffer is written.
+            var allowMemoryBuffer = (_memoryThreshold - count) >= PagedByteBuffer.Length;
+            if (allowMemoryBuffer)
+            {
+                // Buffer content in the MemoryStream if it has capacity.
+                PagedByteBuffer.Add(buffer, offset, count);
+                Debug.Assert(PagedByteBuffer.Length <= _memoryThreshold);
+            }
+            else
+            {
+                // If the MemoryStream is incapable of accomodating the content to be written
+                // spool to disk.
+                EnsureFileStream();
+
+                // Spool memory content to disk and clear in memory buffers. We no longer need to hold on to it
+                PagedByteBuffer.CopyTo(FileStream, clearBuffers: true);
+
+                FileStream.Write(buffer, offset, count);
+            }
+        }
+
+        /// <inheritdoc />
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ThrowArgumentException(buffer, offset, count);
+            ThrowIfDisposed();
+
+            if (_bufferLimit.HasValue && _bufferLimit - BufferedLength < count)
+            {
+                Dispose();
+                throw new IOException("Buffer limit exceeded.");
+            }
+
+            // Allow buffering in memory if we're below the memory threshold once the current buffer is written.
+            var allowMemoryBuffer = (_memoryThreshold - count) >= PagedByteBuffer.Length;
+            if (allowMemoryBuffer)
+            {
+                // Buffer content in the MemoryStream if it has capacity.
+                PagedByteBuffer.Add(buffer, offset, count);
+                Debug.Assert(PagedByteBuffer.Length <= _memoryThreshold);
+            }
+            else
+            {
+                // If the MemoryStream is incapable of accomodating the content to be written
+                // spool to disk.
+                EnsureFileStream();
+
+                // Spool memory content to disk and clear in memory buffers. We no longer need to hold on to it
+                await PagedByteBuffer.CopyToAsync(FileStream, clearBuffers: true, cancellationToken);
```

Comment:
Wait, so this aggregates small writes in memory, dumps the memory to disk when full, but then continues to aggregate in memory? Isn't that going to cause thrashing in the buffer pool if you keep returning them every time?

Reply:
The pool is a synchronized collection, no? And you keep checking out and returning buffers. That's going to cause contention at least. Why not just keep the buffers? Or once you start writing to file, always write to it?

Reply:
> Why not just keep the buffers? 

Because it's more work to keep track of unused buffers. It adds complexity without any known benefits. We can always revisit this in the future once we profile this with large data sets. Most benchmarks today operate on really tiny payloads, so we have no good way of measuring benefits of making the suggested change.